scripts/scale-test/setup_cluster.py

```diff
@@ -0,0 +1,237 @@
+import json
+import logging
+import subprocess
+from argparse import ArgumentParser, Namespace
+from typing import List
+
+from benchmarks import juju
+from benchmarks.models import Cluster, Unit
+from benchmarks.utils import timeit
+
+APPLICATION = "microk8s-node"
+DEFAULT_CHANNEL = "1.24/stable"
+
+logging.basicConfig(format="[%(levelname)s]: %(message)s", level=logging.INFO)
+
+
+@timeit
+def install_microk8s(model, units: List[Unit], channel=DEFAULT_CHANNEL):
+    configure_proxy(units)
+    reboot_and_wait(model)
+    install_snap(channel)
+    update_etc_hosts(units)
+
+
+def update_etc_hosts(units: List[Unit]):
+    """
+    Add entries in /etc/hosts of all units for each node in the cluster.
+    This is needed as nodes's hostnames need to be resolvable in order for
+    the add-node/join process to work.
+    """
+    logging.info("Adding units hostnames on /etc/hosts")
+    for u in units:
+        cmd = f"echo {u.ip}\t{u.instance_id} >> /etc/hosts"
+        juju.run(cmd, app=APPLICATION).check_returncode()
+
+
+@timeit
+def install_snap(channel: str):
+    logging.info("Installing microk8s on all units")
+    for cmd, check_returncode in [
+        (f"snap install microk8s --classic --channel={channel}", True),
+        ("sudo usermod -a -G microk8s ubuntu", True),
+        ("sudo chown -f -R ubuntu ~/.kube", False),
+        ("sudo newgrp microk8s", True),
+        ("microk8s start", True),
+        ("microk8s status --wait-ready", True),
+    ]:
+        resp = juju.run(cmd, app=APPLICATION)
+        if check_returncode:
+            resp.check_returncode()
+
+
+@timeit
+def configure_proxy(units: List[Unit]):
+    logging.info("Configuring proxy settings on units")
+    PROXY = "http://squid.internal:3128"
+    for cmd in [
+        f"echo HTTPS_PROXY={PROXY} >> /etc/environment",
+        f"echo HTTP_PROXY={PROXY} >> /etc/environment",
+        f"echo https_proxy={PROXY} >> /etc/environment",
+        f"echo http_proxy={PROXY} >> /etc/environment",
+    ]:
+        juju.run(cmd, app=APPLICATION).check_returncode()
+
+    # NO_PROXY settings is unit-specific
+    for unit in units:
+        NO_PROXY = f"10.1.0.0/16,10.152.183.0/24,127.0.0.1,{unit.ip},{unit.instance_id},10.246.154.0/24"
```

Comment:
If you go with a bash script triggered by juju you could query the unit IP from within the bash script instead of iterating over each unit.

Reply:
I don't think the bash script option is the way to go, as `juju scp` needs to be executed on a per-node basis. 

Instead, I have found a better way to chain multiple commands to run on all nodes simultaneously:
Before it was 4 `juju run -a` calls
```python
def configure_http_proxy(units: List[Unit], http_proxy: str):
     logging.info("Configuring proxy settings on units")
    for cmd in [
        f"echo HTTPS_PROXY={http_proxy} >> /etc/environment",
        f"echo HTTP_PROXY={http_proxy} >> /etc/environment",
        f"echo https_proxy={http_proxy} >> /etc/environment",
        f"echo http_proxy={http_proxy} >> /etc/environment",
    ]:
        juju.run(cmd, app=APP_NAME).check_returncode()
```

Now just one:
```python
def configure_http_proxy(units: List[Unit], http_proxy: str):
     logging.info("Configuring proxy settings on units")
    proxy_command = ";".join(
        [
            f"echo HTTPS_PROXY={http_proxy} >> /etc/environment",
            f"echo HTTP_PROXY={http_proxy} >> /etc/environment",
            f"echo https_proxy={http_proxy} >> /etc/environment",
            f"echo http_proxy={http_proxy} >> /etc/environment",
        ]
    )
    juju.run(proxy_command, app=APP_NAME).check_returncode()
```

Reply:
Looks better. +1